nipoppy_cli/nipoppy/tabular/dicom_dir_map.py

```diff
@@ -0,0 +1,138 @@
+                f'Session should start with "{BIDS_SESSION_PREFIX}"'
+                f", got {self.session}"
+            )
+        return self
+
+
+class DicomDirMap(BaseTabular):
+    """
+    A dataset's DICOM directory mapping.
+
+    This mapping is used during DICOM reorganization and doughnut generation.
+    """
+
+    # column names
+    col_participant_id = "participant_id"
+    col_session = "session"
+    col_participant_dicom_dir = "participant_dicom_dir"
+
+    index_cols = [col_participant_id, col_session]
+
+    # set the model
+    model = DicomDirMapModel
+
+    _metadata = BaseTabular._metadata + [
+        "col_participant_id",
+        "col_session",
+        "col_participant_dicom_dir",
+        "index_cols",
+        "model",
+    ]
+
+    @classmethod
+    def load_or_generate(
+        cls,
+        manifest: Manifest,
+        fpath_dicom_dir_map: str | Path | None,
+        participant_first: bool,
+        validate: bool = True,
+    ) -> Self:
+        """Load or generate a DicomDirMap instance.
+
+        Parameters
+        ----------
+        manifest : :class:`nipoppy.tabular.manifest.Manifest`
+            Manifest for generating the mapping (not used if ``fpath_dicom_dir_map``
+            is not ``None``).
+        fpath_dicom_dir_map : str | Path | None
+            Path to a custom DICOM directory mapping file. If ``None``,
+            the DICOM directory mapping will be generated from the manifest.
+        participant_first : bool
+            Whether the generated uses ``<PARTICIPANT>/<SESSION>`` order
+            (True) or ``<SESSION>/<PARTICIPANT>`` (False). Not used if
+            ``fpath_dicom_dir_map`` is not ``None``
+        validate : bool, optional
+            Whether to validate (through Pydantic) the created object,
+            by default ``True``
+
+        Returns
+        -------
+        :class:`nipoppy.tabular.dicom_dir_map.DicomDirMap`
+        """
+        # if these is a custom dicom_dir_map, use it
+        if fpath_dicom_dir_map is not None:
+            return cls.load(Path(fpath_dicom_dir_map), validate=validate)
+
+        # else depends on participant_first or no
+        else:
+            data_dicom_dir_map = []
+            for participant, session in manifest.get_participants_sessions():
+                if participant_first:
+                    participant_dicom_dir = f"{participant}/{session}"
+                else:
+                    participant_dicom_dir = f"{session}/{participant}"
+                data_dicom_dir_map.append(
+                    {
+                        cls.col_participant_id: participant,
+                        cls.col_session: session,
+                        cls.col_participant_dicom_dir: participant_dicom_dir,
+                    }
+                )
+            dicom_dir_map = cls(data=data_dicom_dir_map)
+            if validate:
+                dicom_dir_map.validate()
+            return dicom_dir_map
```

Comment:
Not exactly sure how the convention works, but isn't this function suppose to return `Self`?

Reply:
The `Self` (capitalized) type annotation just means it will return a `DicomDirMap` instance, which is what is happening here :)